[Bug]: Provider Crashes

### Terraform Core Version

1.1.7

### AWS Provider Version

4.34.0

### Affected Resource(s)

aws_db_instance

### Expected Behavior

Building an RDS instance from snapshot. Normally this process runs fine with no error.

### Actual Behavior

RDS does not get created and provider just crashes

### Relevant Error/Panic Output Snippet

```shell
Stack trace from the terraform-provider-aws_v4.34.0_x5 plugin:

panic: interface conversion: interface {} is []interface {}, not *schema.Set

goroutine 421 [running]:
github.com/hashicorp/terraform-provider-aws/internal/service/appstream.resourceStackUpdate({0xb0f9b40, 0xc003940120}, 0xc004bd0a80, {0x8fa7980?, 0xc0005b4000?})
github.com/hashicorp/terraform-provider-aws/internal/service/appstream/stack.go:338 +0xbd6
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xb0f9b40?, {0xb0f9b40?, 0xc003940120?}, 0xd?, {0x8fa7980?, 0xc0005b4000?})
github.com/hashicorp/terraform-plugin-sdk/v2@v2.23.0/helper/schema/resource.go:736 +0x87
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000cbec40, {0xb0f9b40, 0xc003940120}, 0xc001f23860, 0xc004bd0780, {0x8fa7980, 0xc0005b4000})
github.com/hashicorp/terraform-plugin-sdk/v2@v2.23.0/helper/schema/resource.go:847 +0x82c
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc00000c6d8, {0xb0f9b40?, 0xc003940000?}, 0xc0049cfbd0)
github.com/hashicorp/terraform-plugin-sdk/v2@v2.23.0/helper/schema/grpc_provider.go:1021 +0xe3c
github.com/hashicorp/terraform-plugin-mux/tf5muxserver.muxServer.ApplyResourceChange({0xc002859440, 0xc0028594a0, {0xc003dd2c20, 0x2, 0x2}, 0xc002859470, 0xc0028832f0, 0xc00396be20, 0xc0028594d0}, {0xb0f9b40, ...}, ...)
github.com/hashicorp/terraform-plugin-mux@v0.7.0/tf5muxserver/mux_server_ApplyResourceChange.go:27 +0x142
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc0016e06e0, {0xb0f9b40?, 0xc00393ced0?}, 0xc0049b6d20)
github.com/hashicorp/terraform-plugin-go@v0.14.0/tfprotov5/tf5server/server.go:818 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0xa041960?, 0xc0016e06e0}, {0xb0f9b40, 0xc00393ced0}, 0xc0049b6cb0, 0x0)
github.com/hashicorp/terraform-plugin-go@v0.14.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc003743860, {0xb0ff440, 0xc0010da1a0}, 0xc004bc7560, 0xc003e2e810, 0x108a5460, 0x0)
google.golang.org/grpc@v1.48.0/server.go:1295 +0xb0b
google.golang.org/grpc.(*Server).handleStream(0xc003743860, {0xb0ff440, 0xc0010da1a0}, 0xc004bc7560, 0x0)
google.golang.org/grpc@v1.48.0/server.go:1636 +0xa1b
google.golang.org/grpc.(*Server).serveStreams.func1.2()
google.golang.org/grpc@v1.48.0/server.go:932 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
google.golang.org/grpc@v1.48.0/server.go:930 +0x28a

Error: The terraform-provider-aws_v4.34.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```


### Terraform Configuration Files

```
resource "aws_db_instance" "oracle-rds" {
    timeouts {
        create = "2h"
        delete = "2h"
    }
    lifecycle {
        ignore_changes = [snapshot_identifier,character_set_name,kms_key_id]
    }
    allocated_storage          = var.snapshot_restore == true ? null : var.rds_allocated_storage
    max_allocated_storage      = 4998
    snapshot_identifier        = var.snapshot_restore == true && var.ignore_snapshot != true ? data.aws_db_snapshot.db_restore_snap[0].id : null # data.aws_db_snapshot.db_restore_snap[0].id

    storage_type               = "gp2"
    engine                     = "oracle-se2"
    instance_class             = var.rds_instance_size
    db_name                    = var.oracle_rds_db_name
    username                   = var.oracle_rds_user
    password                   = var.oracle_rds_pass
    port                       = 1521
    multi_az                   = var.rds_multi_az
    vpc_security_group_ids     = ["${aws_security_group.db_oracle.id}"]
    db_subnet_group_name       = "${aws_db_subnet_group.oracle_subnet_group.name}"
    license_model              = "license-included" # "bring-your-own-license"
    ca_cert_identifier         = "rds-ca-2019"
    maintenance_window         = "wed:05:14-wed:05:44"
    backup_window              = "04:22-04:52"
    backup_retention_period    = var.env != "prod" ? 1 : 3
    skip_final_snapshot        = var.env != "prod" ? true : false
    deletion_protection        = false 
    character_set_name         = "AL32UTF8"
    storage_encrypted          = true
    kms_key_id                 = "${aws_kms_key.rds.arn}"
    option_group_name          = aws_db_option_group.RDSTraxOption.id
    identifier                 = "${var.account_name}-${var.env}-trax-oracle-rds"
    final_snapshot_identifier  = "${var.account_name}-${var.env}-trax-oracle-rds-snapshot-${formatdate("MMDDYYHHmmss", timestamp())}"
    parameter_group_name       = aws_db_parameter_group.db_parameter_group.id

    auto_minor_version_upgrade            = true
    enabled_cloudwatch_logs_exports       = ["alert", "audit", "listener", "trace" ]
    performance_insights_enabled          = true
    performance_insights_retention_period = 7
    monitoring_role_arn                   = var.first_region == true ? aws_iam_role.rds_monitoring_role[0].arn : data.aws_iam_role.rds_monitoring_role[0].arn
    monitoring_interval                   = 60 

}
```

### Steps to Reproduce

Plan works, but once you start the apply it fails right away. 

### Debug Output

_No response_

### Panic Output

_No response_

### Important Factoids

_No response_

### References

_No response_

### Would you like to implement a fix?

_No response_

--- a/.changelog/27257.txt
+++ b/.changelog/27257.txt
@@ -1,0 +1,7 @@
+```release-note:bug
+resource/aws_appstream_stack: Fix panic with `application_settings`.
+```
+
+```release-note:enhancement
+resource/aws_appstream_stack: Add validation for `application_settings`.
+```

--- a/internal/service/appstream/stack.go
+++ b/internal/service/appstream/stack.go
@@ -3,6 +3,7 @@ package appstream
 import (
 	"bytes"
 	"context"
+	"errors"
 	"fmt"
 	"log"
 	"time"
@@ -10,7 +11,9 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/appstream"
 	"github.com/hashicorp/aws-sdk-go-base/v2/awsv1shim/v2/tfawserr"
+	"github.com/hashicorp/go-cty/cty"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/customdiff"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
@@ -19,6 +22,7 @@ import (
 	"github.com/hashicorp/terraform-provider-aws/internal/flex"
 	tftags "github.com/hashicorp/terraform-provider-aws/internal/tags"
 	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
+	"github.com/hashicorp/terraform-provider-aws/internal/verify"
 )
 
 var (
@@ -71,7 +75,7 @@ func ResourceStack() *schema.Resource {
 						"settings_group": {
 							Type:         schema.TypeString,
 							Optional:     true,
-							ValidateFunc: validation.StringLenBetween(1, 100),
+							ValidateFunc: validation.StringLenBetween(0, 100),
 						},
 					},
 				},
@@ -184,7 +188,52 @@ func ResourceStack() *schema.Resource {
 			"tags_all": tftags.TagsSchemaComputed(),
 		},
 
-		CustomizeDiff: verify.SetTagsDiff,
+		CustomizeDiff: customdiff.All(
+			verify.SetTagsDiff,
+			func(_ context.Context, d *schema.ResourceDiff, meta interface{}) error {
+				if d.Id() == "" {
+					return nil
+				}
+
+				rawConfig := d.GetRawConfig()
+				configApplicationSettings := rawConfig.GetAttr("application_settings")
+				if configApplicationSettings.IsKnown() && !configApplicationSettings.IsNull() && configApplicationSettings.LengthInt() > 0 {
+					return nil
+				}
+
+				rawState := d.GetRawState()
+				stateApplicationSettings := rawState.GetAttr("application_settings")
+				if stateApplicationSettings.IsKnown() && !stateApplicationSettings.IsNull() && stateApplicationSettings.LengthInt() > 0 {
+					setting := stateApplicationSettings.Index(cty.NumberIntVal(0))
+					if setting.IsKnown() && !setting.IsNull() {
+						enabled := setting.GetAttr("enabled")
+						if enabled.IsKnown() && !enabled.IsNull() && enabled.True() {
+							// Trigger a diff
+							return d.SetNew("application_settings", []map[string]any{
+								{
+									"enabled":        false,
+									"settings_group": "",
+								},
+							})
+						}
+					}
+				}
+
+				return nil
+			},
+			func(_ context.Context, d *schema.ResourceDiff, meta interface{}) error {
+				_, enabled := d.GetOk("application_settings.0.enabled")
+				v, sg := d.GetOk("application_settings.0.settings_group")
+				log.Print(v)
+
+				if enabled && !sg {
+					return errors.New("application_settings.settings_group must be set when application_settings.enabled is true")
+				} else if !enabled && sg {
+					return errors.New("application_settings.settings_group must not be set when application_settings.enabled is false")
+				}
+				return nil
+			},
+		),
 	}
 }
 
@@ -338,7 +387,7 @@ func resourceStackUpdate(ctx context.Context, d *schema.ResourceData, meta inter
 	}
 
 	if d.HasChange("application_settings") {
-		input.ApplicationSettings = expandApplicationSettings(d.Get("application_settings").(*schema.Set).List())
+		input.ApplicationSettings = expandApplicationSettings(d.Get("application_settings").([]interface{}))
 	}
 
 	if d.HasChange("description") {
@@ -477,38 +526,20 @@ func flattenAccessEndpoints(apiObjects []*appstream.AccessEndpoint) []map[string
 	return tfList
 }
 
-func expandApplicationSetting(tfMap map[string]interface{}) *appstream.ApplicationSettings {
-	if tfMap == nil {
-		return nil
+func expandApplicationSettings(tfList []interface{}) *appstream.ApplicationSettings {
+	if len(tfList) == 0 {
+		return &appstream.ApplicationSettings{
+			Enabled: aws.Bool(false),
+		}
 	}
 
-	apiObject := &appstream.ApplicationSettings{}
+	tfMap := tfList[0].(map[string]interface{})
 
-	if v, ok := tfMap["enabled"]; ok {
-		apiObject.Enabled = aws.Bool(v.(bool))
+	apiObject := &appstream.ApplicationSettings{
+		Enabled: aws.Bool(tfMap["enabled"].(bool)),
 	}
 	if v, ok := tfMap["settings_group"]; ok {
 		apiObject.SettingsGroup = aws.String(v.(string))
-	}
-
-	return apiObject
-}
-
-func expandApplicationSettings(tfList []interface{}) *appstream.ApplicationSettings {
-	if len(tfList) == 0 {
-		return nil
-	}
-
-	var apiObject *appstream.ApplicationSettings
-
-	for _, tfMapRaw := range tfList {
-		tfMap, ok := tfMapRaw.(map[string]interface{})
-
-		if !ok {
-			continue
-		}
-
-		apiObject = expandApplicationSetting(tfMap)
 	}
 
 	return apiObject
@@ -519,11 +550,10 @@ func flattenApplicationSetting(apiObject *appstream.ApplicationSettingsResponse)
 		return nil
 	}
 
-	tfMap := map[string]interface{}{}
-	tfMap["enabled"] = aws.BoolValue(apiObject.Enabled)
-	tfMap["settings_group"] = aws.StringValue(apiObject.SettingsGroup)
-
-	return tfMap
+	return map[string]interface{}{
+		"enabled":        aws.BoolValue(apiObject.Enabled),
+		"settings_group": aws.StringValue(apiObject.SettingsGroup),
+	}
 }
 
 func flattenApplicationSettings(apiObject *appstream.ApplicationSettingsResponse) []interface{} {

--- a/internal/service/appstream/stack.go
+++ b/internal/service/appstream/stack.go
@@ -66,11 +66,12 @@ func ResourceStack() *schema.Resource {
 					Schema: map[string]*schema.Schema{
 						"enabled": {
 							Type:     schema.TypeBool,
-							Optional: true,
+							Required: true,
 						},
 						"settings_group": {
-							Type:     schema.TypeString,
-							Optional: true,
+							Type:         schema.TypeString,
+							Optional:     true,
+							ValidateFunc: validation.StringLenBetween(1, 100),
 						},
 					},
 				},
@@ -182,6 +183,8 @@ func ResourceStack() *schema.Resource {
 			"tags":     tftags.TagsSchemaForceNew(),
 			"tags_all": tftags.TagsSchemaComputed(),
 		},
+
+		CustomizeDiff: verify.SetTagsDiff,
 	}
 }
 

--- a/internal/service/appstream/stack.go
+++ b/internal/service/appstream/stack.go
@@ -336,14 +336,14 @@ func resourceStackRead(ctx context.Context, d *schema.ResourceData, meta interfa
 			return diag.FromErr(fmt.Errorf("error setting `%s` for AppStream Stack (%s): %w", "access_endpoints", d.Id(), err))
 		}
 		if err = d.Set("application_settings", flattenApplicationSettings(v.ApplicationSettings)); err != nil {
-			return diag.FromErr(fmt.Errorf("error setting `%s` for AppStream Stack (%s): %w", "user_settings", d.Id(), err))
+			return diag.FromErr(fmt.Errorf("error setting `%s` for AppStream Stack (%s): %w", "application_settings", d.Id(), err))
 		}
 		d.Set("arn", v.Arn)
 		d.Set("created_time", aws.TimeValue(v.CreatedTime).Format(time.RFC3339))
 		d.Set("description", v.Description)
 		d.Set("display_name", v.DisplayName)
 		if err = d.Set("embed_host_domains", flex.FlattenStringList(v.EmbedHostDomains)); err != nil {
-			return diag.FromErr(fmt.Errorf("error setting `%s` for AppStream Stack (%s): %w", "user_settings", d.Id(), err))
+			return diag.FromErr(fmt.Errorf("error setting `%s` for AppStream Stack (%s): %w", "embed_host_domains", d.Id(), err))
 		}
 		d.Set("feedback_url", v.FeedbackURL)
 		d.Set("name", v.Name)

--- a/internal/service/appstream/stack_test.go
+++ b/internal/service/appstream/stack_test.go
@@ -136,6 +136,132 @@ func TestAccAppStreamStack_complete(t *testing.T) {
 	})
 }
 
+func TestAccAppStreamStack_applicationSettings_basic(t *testing.T) {
+	var stackOutput appstream.Stack
+	resourceName := "aws_appstream_stack.test"
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+	settingsGroup := "group"
+	settingsGroupUpdated := "group-updated"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckStackDestroy,
+		ErrorCheck:               acctest.ErrorCheck(t, appstream.EndpointsID),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccStackConfig_applicationSettings(rName, true, settingsGroup),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckStackExists(resourceName, &stackOutput),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.enabled", "true"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.settings_group", settingsGroup),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
+				Config: testAccStackConfig_applicationSettings(rName, true, settingsGroupUpdated),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckStackExists(resourceName, &stackOutput),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.enabled", "true"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.settings_group", settingsGroupUpdated),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+			{
+				Config: testAccStackConfig_applicationSettings(rName, false, ""),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckStackExists(resourceName, &stackOutput),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.enabled", "false"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.settings_group", ""),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAccAppStreamStack_applicationSettings_removeFromEnabled(t *testing.T) {
+	var stackOutput appstream.Stack
+	resourceName := "aws_appstream_stack.test"
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+	settingsGroup := "group"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckStackDestroy,
+		ErrorCheck:               acctest.ErrorCheck(t, appstream.EndpointsID),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccStackConfig_applicationSettings(rName, true, settingsGroup),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckStackExists(resourceName, &stackOutput),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.enabled", "true"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.settings_group", settingsGroup),
+				),
+			},
+			{
+				Config: testAccStackConfig_applicationSettingsRemoved(rName),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckStackExists(resourceName, &stackOutput),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.enabled", "false"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.settings_group", ""),
+				),
+			},
+			{
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateVerify: true,
+			},
+		},
+	})
+}
+
+func TestAccAppStreamStack_applicationSettings_removeFromDisabled(t *testing.T) {
+	var stackOutput appstream.Stack
+	resourceName := "aws_appstream_stack.test"
+	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                 func() { acctest.PreCheck(t) },
+		ProtoV5ProviderFactories: acctest.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccCheckStackDestroy,
+		ErrorCheck:               acctest.ErrorCheck(t, appstream.EndpointsID),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccStackConfig_applicationSettingsDisabled(rName),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckStackExists(resourceName, &stackOutput),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.enabled", "false"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.settings_group", ""),
+				),
+			},
+			{
+				Config:   testAccStackConfig_applicationSettingsRemoved(rName),
+				PlanOnly: true,
+			},
+		},
+	})
+}
+
 func TestAccAppStreamStack_withTags(t *testing.T) {
 	var stackOutput appstream.Stack
 	resourceName := "aws_appstream_stack.test"
@@ -277,6 +403,39 @@ resource "aws_appstream_stack" "test" {
   }
 }
 `, name, description)
+}
+
+func testAccStackConfig_applicationSettings(name string, enabled bool, settingsGroupName string) string {
+	return fmt.Sprintf(`
+resource "aws_appstream_stack" "test" {
+  name = %[1]q
+
+  application_settings {
+    enabled        = %[2]t
+    settings_group = %[3]q
+  }
+}
+`, name, enabled, settingsGroupName)
+}
+
+func testAccStackConfig_applicationSettingsDisabled(name string) string {
+	return fmt.Sprintf(`
+resource "aws_appstream_stack" "test" {
+  name = %[1]q
+
+  application_settings {
+    enabled = false
+  }
+}
+`, name)
+}
+
+func testAccStackConfig_applicationSettingsRemoved(name string) string {
+	return fmt.Sprintf(`
+resource "aws_appstream_stack" "test" {
+  name = %[1]q
+}
+`, name)
 }
 
 func testAccStackConfig_tags(name, description string) string {

--- a/internal/service/appstream/stack_test.go
+++ b/internal/service/appstream/stack_test.go
@@ -28,10 +28,23 @@ func TestAccAppStreamStack_basic(t *testing.T) {
 		Steps: []resource.TestStep{
 			{
 				Config: testAccStackConfig_basic(rName),
-				Check: resource.ComposeTestCheckFunc(
+				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccCheckStackExists(resourceName, &stackOutput),
 					resource.TestCheckResourceAttr(resourceName, "name", rName),
 					acctest.CheckResourceAttrRFC3339(resourceName, "created_time"),
+					resource.TestCheckResourceAttr(resourceName, "access_endpoints.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.enabled", "false"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.settings_group", ""),
+					resource.TestCheckResourceAttr(resourceName, "description", ""),
+					resource.TestCheckResourceAttr(resourceName, "display_name", ""),
+					resource.TestCheckResourceAttr(resourceName, "embed_host_domains.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "feedback_url", ""),
+					resource.TestCheckResourceAttr(resourceName, "redirect_url", ""),
+					resource.TestCheckResourceAttr(resourceName, "storage_connectors.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "user_settings.#", "7"),
+					resource.TestCheckResourceAttr(resourceName, "tags.%", "0"),
+					resource.TestCheckResourceAttr(resourceName, "tags_all.%", "0"),
 				),
 			},
 			{
@@ -56,7 +69,7 @@ func TestAccAppStreamStack_disappears(t *testing.T) {
 		Steps: []resource.TestStep{
 			{
 				Config: testAccStackConfig_basic(rName),
-				Check: resource.ComposeTestCheckFunc(
+				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccCheckStackExists(resourceName, &stackOutput),
 					acctest.CheckResourceDisappears(acctest.Provider, tfappstream.ResourceStack(), resourceName),
 				),
@@ -81,16 +94,30 @@ func TestAccAppStreamStack_complete(t *testing.T) {
 		Steps: []resource.TestStep{
 			{
 				Config: testAccStackConfig_complete(rName, description),
-				Check: resource.ComposeTestCheckFunc(
+				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccCheckStackExists(resourceName, &stackOutput),
 					resource.TestCheckResourceAttr(resourceName, "name", rName),
 					acctest.CheckResourceAttrRFC3339(resourceName, "created_time"),
 					resource.TestCheckResourceAttr(resourceName, "description", description),
+					resource.TestCheckResourceAttr(resourceName, "embed_host_domains.#", "2"),
+					resource.TestCheckTypeSetElemAttr(resourceName, "embed_host_domains.*", "example.com"),
+					resource.TestCheckTypeSetElemAttr(resourceName, "embed_host_domains.*", "subdomain.example.com"),
+					resource.TestCheckResourceAttr(resourceName, "access_endpoints.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.enabled", "true"),
+					resource.TestCheckResourceAttr(resourceName, "application_settings.0.settings_group", "SettingsGroup"),
+					resource.TestCheckResourceAttr(resourceName, "display_name", ""),
+					resource.TestCheckResourceAttr(resourceName, "feedback_url", ""),
+					resource.TestCheckResourceAttr(resourceName, "redirect_url", ""),
+					resource.TestCheckResourceAttr(resourceName, "storage_connectors.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "user_settings.#", "4"),
+					resource.TestCheckResourceAttr(resourceName, "tags.%", "0"),
+					resource.TestCheckResourceAttr(resourceName, "tags_all.%", "0"),
 				),
 			},
 			{
 				Config: testAccStackConfig_complete(rName, descriptionUpdated),
-				Check: resource.ComposeTestCheckFunc(
+				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccCheckStackExists(resourceName, &stackOutput),
 					resource.TestCheckResourceAttr(resourceName, "name", rName),
 					acctest.CheckResourceAttrRFC3339(resourceName, "created_time"),
@@ -124,7 +151,7 @@ func TestAccAppStreamStack_withTags(t *testing.T) {
 		Steps: []resource.TestStep{
 			{
 				Config: testAccStackConfig_complete(rName, description),
-				Check: resource.ComposeTestCheckFunc(
+				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccCheckStackExists(resourceName, &stackOutput),
 					resource.TestCheckResourceAttr(resourceName, "name", rName),
 					acctest.CheckResourceAttrRFC3339(resourceName, "created_time"),
@@ -133,7 +160,7 @@ func TestAccAppStreamStack_withTags(t *testing.T) {
 			},
 			{
 				Config: testAccStackConfig_tags(rName, descriptionUpdated),
-				Check: resource.ComposeTestCheckFunc(
+				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccCheckStackExists(resourceName, &stackOutput),
 					resource.TestCheckResourceAttr(resourceName, "name", rName),
 					acctest.CheckResourceAttrRFC3339(resourceName, "created_time"),

--- a/website/docs/r/appstream_stack.html.markdown
+++ b/website/docs/r/appstream_stack.html.markdown
@@ -60,31 +60,45 @@ The following arguments are required:
 
 The following arguments are optional:
 
-* `access_endpoints` - (Optional) Set of configuration blocks defining the interface VPC endpoints. Users of the stack can connect to AppStream 2.0 only through the specified endpoints. See below.
+* `access_endpoints` - (Optional) Set of configuration blocks defining the interface VPC endpoints. Users of the stack can connect to AppStream 2.0 only through the specified endpoints.
+  See [`access_endpoints`](#access_endpoints) below.
 * `application_settings` - (Optional) Settings for application settings persistence.
+  See [`application_settings`](#application_settings) below.
 * `description` - (Optional) Description for the AppStream stack.
 * `display_name` - (Optional) Stack name to display.
 * `embed_host_domains` - (Optional) Domains where AppStream 2.0 streaming sessions can be embedded in an iframe. You must approve the domains that you want to host embedded AppStream 2.0 streaming sessions.
 * `feedback_url` - (Optional) URL that users are redirected to after they click the Send Feedback link. If no URL is specified, no Send Feedback link is displayed. .
 * `redirect_url` - (Optional) URL that users are redirected to after their streaming session ends.
-* `storage_connectors` - (Optional) Configuration block for the storage connectors to enable. See below.
-* `user_settings` - (Optional) Configuration block for the actions that are enabled or disabled for users during their streaming sessions. By default, these actions are enabled. See below.
+* `storage_connectors` - (Optional) Configuration block for the storage connectors to enable.
+  See [`storage_connectors`](#storage_connectors) below.
+* `user_settings` - (Optional) Configuration block for the actions that are enabled or disabled for users during their streaming sessions. By default, these actions are enabled.
+  See [`user_settings`](#user_settings) below.
 
 ### `access_endpoints`
 
-* `endpoint_type` - (Required) Type of the interface endpoint. See the [`AccessEndpoint` AWS API documentation](https://docs.aws.amazon.com/appstream2/latest/APIReference/API_AccessEndpoint.html) for valid values.
+* `endpoint_type` - (Required) Type of the interface endpoint.
+  See the [`AccessEndpoint` AWS API documentation](https://docs.aws.amazon.com/appstream2/latest/APIReference/API_AccessEndpoint.html) for valid values.
 * `vpce_id` - (Optional) ID of the VPC in which the interface endpoint is used.
+
+### `application_settings`
+
+* `enabled` - (Required) Whether application settings should be persisted.
+* `settings_group` - (Optional) Name of the settings group.
+  Required when `enabled` is `true`.
 
 ### `storage_connectors`
 
-* `connector_type` - (Required) Type of storage connector. Valid values are: `HOMEFOLDERS`, `GOOGLE_DRIVE`, `ONE_DRIVE`.
+* `connector_type` - (Required) Type of storage connector.
+  Valid values are `HOMEFOLDERS`, `GOOGLE_DRIVE`, or `ONE_DRIVE`.
 * `domains` - (Optional) Names of the domains for the account.
 * `resource_identifier` - (Optional) ARN of the storage connector.
 
 ### `user_settings`
 
-* `action` - (Required) Action that is enabled or disabled. Valid values are: `CLIPBOARD_COPY_FROM_LOCAL_DEVICE`,  `CLIPBOARD_COPY_TO_LOCAL_DEVICE`, `FILE_UPLOAD`, `FILE_DOWNLOAD`, `PRINTING_TO_LOCAL_DEVICE`, `DOMAIN_PASSWORD_SIGNIN`, `DOMAIN_SMART_CARD_SIGNIN`.
-* `permission` - (Required) Whether the action is enabled or disabled. Valid values are: `ENABLED`, `DISABLED`.
+* `action` - (Required) Action that is enabled or disabled.
+  Valid values are `CLIPBOARD_COPY_FROM_LOCAL_DEVICE`,  `CLIPBOARD_COPY_TO_LOCAL_DEVICE`, `FILE_UPLOAD`, `FILE_DOWNLOAD`, `PRINTING_TO_LOCAL_DEVICE`, `DOMAIN_PASSWORD_SIGNIN`, or `DOMAIN_SMART_CARD_SIGNIN`.
+* `permission` - (Required) Whether the action is enabled or disabled.
+  Valid values are `ENABLED` or `DISABLED`.
 
 ## Attributes Reference
 

--- a/website/docs/r/appstream_stack.html.markdown
+++ b/website/docs/r/appstream_stack.html.markdown
@@ -85,6 +85,7 @@ The following arguments are optional:
 * `enabled` - (Required) Whether application settings should be persisted.
 * `settings_group` - (Optional) Name of the settings group.
   Required when `enabled` is `true`.
+  Can be up to 100 characters.
 
 ### `storage_connectors`
 

